physical/alibaba/alibaba.go

```diff
@@ -0,0 +1,223 @@
+package alibaba
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/aliyun/aliyun-oss-go-sdk/oss"
+	log "github.com/hashicorp/go-hclog"
+
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/errwrap"
+	"github.com/hashicorp/vault/physical"
+)
+
+// Verify AlibabaBackend satisfies the correct interfaces
+var _ physical.Backend = (*AlibabaBackend)(nil)
+
+// AlibabaBackend is a physical backend that stores data
+// within an Alibaba OSS bucket.
+type AlibabaBackend struct {
+	bucket     string
+	client     *oss.Client
+	logger     log.Logger
+	permitPool *physical.PermitPool
+}
+
+// NewAlibabaBackend constructs an OSS backend using a pre-existing
+// bucket. Credentials can be provided to the backend, sourced
+// from the environment.
+func NewAlibabaBackend(conf map[string]string, logger log.Logger) (physical.Backend, error) {
+	endpoint := os.Getenv("ALIBABA_OSS_ENDPOINT")
+	if endpoint == "" {
+		endpoint = conf["endpoint"]
+		if endpoint == "" {
+			return nil, fmt.Errorf("'endpoint' must be set")
+		}
+	}
+
+	bucket := os.Getenv("ALIBABA_OSS_BUCKET")
+	if bucket == "" {
+		bucket = conf["bucket"]
+		if bucket == "" {
+			return nil, fmt.Errorf("'bucket' must be set")
+		}
+	}
+
+	accessKeyID := os.Getenv("ALIBABA_ACCESS_KEY_ID")
+	if accessKeyID == "" {
+		accessKeyID = conf["access_key_id"]
+		if accessKeyID == "" {
+			return nil, fmt.Errorf("'access_key_id' must be set")
+		}
+	}
+
+	accessKeySecret := os.Getenv("ALIBABA_ACCESS_KEY_SECRET")
```

Comment:
Can you rename this to `ALICLOUD_ACCESS_KEY`?

Reply:
Err, I mean `ALICLOUD_SECRET_KEY`.

Reply:
FWIW, the AWS ones are called:
- AWS_ACCESS_KEY_ID
- AWS_SECRET_ACCESS_KEY

AFAICT, AliCloud doesn't support credentials via env vars, but when you do download creds from their UI and get a little text file containing them, AliCloud calls them:
- AccessKeyId
- AccessKeySecret

Could I propose ALICLOUD_ACCESS_KEY_ID and ALICLOUD_ACCESS_KEY_SECRET?

Reply:
Whatever's decided, I'll use it in my code too.

Reply:
I would vote for ALICLOUD_ACCESS_KEY_ID and ALICLOUD_ACCESS_KEY_SECRET as well, and I will come back with the documentation in the second part of this week.

Reply:
This is mainly suggested to match Terraform and Packer.  Even if these var names are not perfect, I would like to consistency across products for the chosen values.

https://www.terraform.io/docs/providers/alicloud/index.html#access_key
https://www.packer.io/docs/builders/alicloud-ecs.html#access_key

Reply:
LGTM. :-) Thanks @bonifaido!